File: 1046-last-stone-weight/1046-last-stone-weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        heap = [-s for s in stones]
        heapq.heapify(heap)
        
        while len(heap) > 1:
            stone1 = heapq.heappop(heap)
            stone2 = heapq.heappop(heap)
            
            if stone1 != stone2:
                heapq.heappush(heap, stone1 - stone2)
        
        if heap:
            return -1 * heap[0]
        return 0
                

File: 1046-last-stone-weight/1046-last-stone-weight.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> heapq = new PriorityQueue<Integer>(stones.length);
        for (int n : stones) {
            heapq.add(-n);
        }
        
        while (heapq.size() > 1) {
            int y = -1 * heapq.poll();
            int x = -1 * heapq.poll();
            if (x != y) {
                heapq.add(x - y);
            }
        }
        
        if (heapq.size() == 0) return 0;
        return -1 * heapq.poll();
    }
}